core/service.go

```diff
@@ -28,117 +32,200 @@ import (
 		return nil, err
 	}
 
-	return &Service{
+	s := &Service{
 		rt:      rt,
-		b:       b,
-		msgRec:  msgRec,
-		msgSend: msgSend,
-	}, nil
+		bs:      bs,
+		bsRec:   bsRec,
+		p2pRec:  p2pRec,
+		p2pSend: p2pSend,
+	}
+
+	return s, nil
 }
 
-// Start begins the service. This begins watching the message channel for new block or transaction messages.
+// Start starts the core service
 func (s *Service) Start() error {
-	e := make(chan error)
-	go s.start(e)
-	return <-e
-}
 
-func (s *Service) start(e chan error) {
-	e <- nil
+	// start receiving blocks from BABE session
+	go s.startReceivingBlocks()
 
-	for {
-		msg, ok := <-s.msgRec
-		if !ok {
-			log.Warn("core service message watcher", "error", "channel closed")
-			break
-		}
+	// start receiving messages from p2p service
+	go s.startReceivingMessages()
 
-		msgType := msg.GetType()
-
-		switch msgType {
-		case p2p.TransactionMsgType:
-			err := s.ProcessTransactionMessage(msg)
-			if err != nil {
-				log.Error("core service", "error", err)
-				e <- err
-			}
-			e <- nil
-		case p2p.BlockAnnounceMsgType:
-			// get extrinsics by sending BlockRequest message
-			// process block announce message
-		case p2p.BlockResponseMsgType:
-			err := s.ProcessBlockResponseMessage(msg)
-			if err != nil {
-				log.Error("core service", "error", err)
-				e <- err
-			}
-			e <- nil
-		default:
-			log.Error("core service", "error", "got unsupported message type")
-		}
-	}
+	return nil
 }
 
+// Stop stops the core service
 func (s *Service) Stop() error {
+
+	// stop runtime
 	if s.rt != nil {
 		s.rt.Stop()
 	}
-	if s.msgSend != nil {
-		close(s.msgSend)
+
+	// close p2pSend channel
+	if s.p2pSend != nil {
+		close(s.p2pSend)
 	}
+
 	return nil
 }
 
+// StorageRoot returns the hash of the runtime storage root
 func (s *Service) StorageRoot() (common.Hash, error) {
 	return s.rt.StorageRoot()
 }
 
-// ProcessTransactionMessage attempts to validates the transaction
-// if it is validated, it is added to the transaction pool of the BABE session
-func (s *Service) ProcessTransactionMessage(msg p2p.Message) error {
-	extrinsics := msg.(*p2p.TransactionMessage).Extrinsics
-
-	for _, extrinsic := range extrinsics {
-		extrinsic := extrinsic // pin
+// startReceivingBlocks starts receiving blocks from the BABE session
+func (s *Service) startReceivingBlocks() {
+	for {
+		block, ok := <-s.bsRec
+		if !ok {
+			log.Debug("receive channel closed")
+			return // exit
+		}
+		err := s.handleBlock(block)
+		if err != nil {
+			log.Debug("handle block", "err", err)
+		}
+	}
+}
 
-		validity, err := s.validateTransaction(extrinsic)
+// startReceivingMessages starts receiving messages from the p2p service
+func (s *Service) startReceivingMessages() {
+	for {
+		msg, ok := <-s.p2pRec
+		if !ok {
+			log.Debug("receive channel closed")
+			return // exit
+		}
+		err := s.handleMessage(msg)
 		if err != nil {
-			log.Error("ProcessTransaction", "error", err)
-			return err
+			log.Debug("handle message", "err", err)
```

Comment:
just listing out the function thats called, please update this based on our discussions on your other PR.

Should also be error.

Reply:
Fixed.